extensions/sovity-messenger/src/main/java/de/sovity/edc/extension/messenger/api/MessageHandlerRegistry.java

```diff
@@ -0,0 +1,78 @@
+/*
+ *  Copyright (c) 2024 sovity GmbH
+ *
+ *  This program and the accompanying materials are made available under the
+ *  terms of the Apache License, Version 2.0 which is available at
+ *  https://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  SPDX-License-Identifier: Apache-2.0
+ *
+ *  Contributors:
+ *       sovity GmbH - initial API and implementation
+ */
+
+package de.sovity.edc.extension.messenger.api;
+
+import lombok.val;
+import org.eclipse.edc.spi.EdcException;
+
+import java.lang.reflect.InvocationTargetException;
+import java.util.function.Function;
+
+/**
+ * The component where to register message handlers when using the {@link SovityMessenger}.
+ */
+public interface MessageHandlerRegistry {
+    /**
+     * Register a handler to process a sovity message.
+     * Prefer using {@link MessageHandlerRegistry#register(Class, Function, Object[])} for a typesafe, non-string-based registration.
+     */
+    @SuppressWarnings("unchecked")
+    default <IN, OUT> void register(String type, Function<IN, OUT> handler, IN... reified) {
+        register(getClassOf(reified), type, handler);
+    }
+
+    /**
+     * Register a handler to process a sovity message.
+     *
+     * @param outgoingMessage The outgoing message type to register
+     * @param handler         The function to process this type of message.
+     * @param reified         Convenience to avoid repeating the incoming message type;
+     * @param <IN>            Incoming message type.
+     * @param <OUT>           Outgoing message type you send with {@link SovityMessenger#send(Class, String, SovityMessage)}.
+     */
+    @SuppressWarnings("unchecked")
+    default <IN, OUT> void register(Class<? extends SovityMessage> outgoingMessage, Function<IN, OUT> handler, IN... reified) {
```

Comment:
the reason why you needed the casts at the parent level is because you moved the "extends Sovity Message" inside a single parameter, right?

```java
default <IN extends SovityMessage, OUT> void register(Class<IN> outgoingMessage, Function<IN, OUT> handler) {
```